### PR TITLE
*: AnalyzeV2 supports processing special global indexes

### LINKED ISSUE #54233: Statistics cannot be generated correctly for a global index that has an expression index
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
set tidb_enable_global_index=true;
drop table if exists t;
CREATE TABLE `t` (
  `a` int(11) DEFAULT NULL,
  `b` int(11) DEFAULT NULL,
  `c` int(11) DEFAULT NULL,
  UNIQUE KEY `idx` ((`c` + 1)),
  UNIQUE KEY `idx1` (`c`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
PARTITION BY HASH (`b`) PARTITIONS 2;
insert into t values (1,1,1),(2,2,2),(3,3,3),(4,4,4),(5,5,5);
analyze table t;
explain select c+1 from t where (c+1)>3;
```

### 2. What did you expect to see? (Required)

```
mysql> explain select c+1 from t where (c+1)>3;
+------------------------+---------+-----------+------------------------------+--------------------------------------------------------------------------+
| id                     | estRows | task      | access object                | operator info                                                            |
+------------------------+---------+-----------+------------------------------+--------------------------------------------------------------------------+
| IndexReader_6          | 3.00    | root      |                              | index:IndexRangeScan_5                                                   |
| └─IndexRangeScan_5     | 3.00    | cop[tikv] | table:t, index:idx(`c` + 1) | range:(3,+inf], keep order:false, stats:partial[_v$_idx_0:unInitialized] |
+------------------------+---------+-----------+------------------------------+--------------------------------------------------------------------------+
2 rows in set (0.01 sec)
```

### 3. What did you see instead (Required)

```
mysql> explain select c+1 from t where (c+1)>3;
+------------------------+---------+-----------+-----------------------------+--------------------------------------------------------------------------+
| id                     | estRows | task      | access object               | operator info                                                            |
+------------------------+---------+-----------+-----------------------------+--------------------------------------------------------------------------+
| IndexReader_7          | 2.08    | root      | partition:all               | index:IndexRangeScan_5                                                   |
| └─IndexRangeScan_5     | 2.08    | cop[tikv] | table:t, index:idx(`c` + 1) | range:(3,+inf], keep order:false, stats:partial[_v$_idx_0:unInitialized] |
+------------------------+---------+-----------+-----------------------------+--------------------------------------------------------------------------+
2 rows in set (0.01 sec)
```

The estRows are never changed, no matter what conditions are used to query, e.g. `explain select c+1 from t where (c+1)>0;`

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #54233

Problem Summary:

Supports processing global index in analyze version V2. Because the data of the global index is not scattered in every partitions, only one analyze is required for the whole table.

Because we use an independent index analyze task, there will be results of the same version when saving, which needs to be processed like mv index.

After the analysis is complete, the global index will only have global stats.

AutoAnalyze will not process global index, user should trigger it manually.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/planbuilder.go [hunk 1]
```diff
@@ -2385,6 +2391,37 @@ func (b *PlanBuilder) buildAnalyzeFullSamplingTask(
...
+			if !handleutil.IsSpecialGlobalIndex(idx, tbl.TableInfo) {
+				allSpecialGlobalIndex = false
+				break
+			}
+			if !isAnalyzeTable {
+				return errors.NewNoStackErrorf("Analyze global index '%s' can't work with analyze specified partitions", idx.Name.O)
+			}
+		}
+	}
+
```
Comment (Rustin170506) [inline pkg/planner/core/planbuilder.go:2424]: How about moving this check to a separate function?

**Review by Rustin170506 — COMMENTED** [review]
Thanks!

File: pkg/planner/core/planbuilder.go (hunk 1 above)
Comment (Defined2014) [reply pkg/planner/core/planbuilder.go]: Done, PTAL

File: pkg/statistics/handle/util/util.go [hunk 2]
```diff
@@ -309,3 +311,19 @@ type JSONPredicateColumn struct {
 	LastAnalyzedAt *string `json:"last_analyzed_at"`
 	ID             int64   `json:"id"`
 }
+
+// IsSpecialGlobalIndex checks a index is a special global index or not.
+// A special global index is one that is a global index and has virtual generated columns or prefix columns.
+func IsSpecialGlobalIndex(idx *model.IndexInfo, tblInfo *model.TableInfo) bool {
```
Comment (Rustin170506) [inline pkg/statistics/handle/util/util.go:317]: Better to add some unit tests for it.

File: pkg/planner/core/planbuilder.go [hunk 3]
```diff
@@ -2414,51 +2464,76 @@ func (b *PlanBuilder) buildAnalyzeFullSamplingTask(
...
+					IndexInfo:   indexInfo,
+					TblInfo:     tbl.TableInfo,
+					AnalyzeInfo: info,
+				}
+				analyzePlan.IdxTasks = append(analyzePlan.IdxTasks, newIdxTask)
+			}
+		}
+	}
+
+	if isAnalyzeTable {
```
Comment (Rustin170506) [inline pkg/planner/core/planbuilder.go:2524]: Could you please add some comments here to explain why we need this branch here?

File: tests/integrationtest/t/globalindex/analyze.test [hunk 4]
```diff
@@ -0,0 +1,70 @@
+set tidb_enable_global_index=true;
```
Comment (Rustin170506) [inline tests/integrationtest/t/globalindex/analyze.test:1]: Could you please provide a case to cover the analysis of multiple global indexes here?

**Review by Rustin170506 — APPROVED** [review]
rest LGTM

/hold

Feel free to unhold it after you fix them. Or feel free to cc me to look at it again.

**Review by winoros — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**